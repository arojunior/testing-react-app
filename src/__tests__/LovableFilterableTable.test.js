import React from 'react'
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import serializer from 'jest-serializer-enzyme'
import LovableFilterableTable from '../LovableFilterableTable'
import { tableSchema } from '../App'

import fs from 'fs'
import path from 'path'

const SAMPLE_RESPONSE_FILE = path.join(__dirname, '../sample-response.json')

const generateItems = () => {
  const response = fs.readFileSync(SAMPLE_RESPONSE_FILE)
  const json = JSON.parse(response)

  return json.slice(0, 30).map(item => ({
    ...item,
    isLoved: false
  }))
}

describe('LovableFilterableTable', () => {
  expect.addSnapshotSerializer(serializer)

  let wrapper

  describe('when given empty `items`', () => {
    const items = []

    beforeEach(() => {
      wrapper = shallow(
        <LovableFilterableTable items={items} schema={tableSchema} />
      )
    })

    it('should still render search box', () => {
      expect(wrapper.find('input').exists()).toBe(true)
    })

    it('should have no table rows', () => {
      expect(wrapper.find('tbody > tr').exists()).toBe(false)
    })
  })

  describe('when given some `items`', () => {
    // Presence in this array does not indicate endorsement
    const items = [
      { id: 1, name: 'Bitcoin' },
      { id: 2, name: 'Ethereum' },
      { id: 3, name: 'Litecoin' }
    ]

    beforeEach(() => {
      wrapper = shallow(
        <LovableFilterableTable items={items} schema={tableSchema} />
      )
    })

    it('should render corresponding number of table rows', () => {
      expect(wrapper.find('tbody > tr').length).toEqual(3)
    })

    it('should include the title of each item', () => {
      items.forEach(item => {
        expect(
          wrapper.containsMatchingElement(
            <td>
              {item.name}
            </td>
          )
        ).toBe(true)
      })
    })
  })

  describe('user enters search query', () => {
    let items

    beforeEach(() => {
      items = generateItems()
      wrapper = shallow(
        <LovableFilterableTable items={items} schema={tableSchema} />
      )

      const searchBox = wrapper.find('input')
      searchBox.simulate('change', { target: { value: 'coin' } })
    })

    it('should render a subset of matching `items`', () => {
      const matching = items.filter(i => i.name.match(/coin/i))

      matching.forEach(match => {
        expect(
          wrapper.containsMatchingElement(
            <td>
              {match.name}
            </td>
          )
        ).toBe(true)
      })
    })

    it("should not render the `items` that don't match", () => {
      const notMatching = items.filter(i => !i.name.match(/coin/i))

      notMatching.forEach(match => {
        expect(
          wrapper.containsMatchingElement(
            <td>
              {match.name}
            </td>
          )
        ).toBe(false)
      })
    })

    it('testing out snapshots', () => {
      const item = items[0]
      expect(item).toMatchSnapshot()
    })

    it('should render a subset of matching `items`', () => {
      expect(wrapper.find('tbody').first().html()).toMatchSnapshot()
    })

    it('should filter items', () => {
      expect(
        wrapper.find('tbody > tr > .item-name').map(i => i.html())
      ).toMatchSnapshot()
    })
    it('should filter items (with serializer!)', () => {
      expect(wrapper.find('tbody')).toMatchSnapshot()
    })
  })
  describe('user clears search query', () => {
    let items
    beforeEach(() => {
      items = generateItems()
      wrapper = shallow(
        <LovableFilterableTable items={items} schema={tableSchema} />
      )      
      const searchBox = wrapper.find('input')
      searchBox.simulate('change', { target: { value: '' } })
    })

    it('should render all the items again', () => {
      expect(wrapper.find('tbody > tr').length).toEqual(items.length)
    })
  })
})
