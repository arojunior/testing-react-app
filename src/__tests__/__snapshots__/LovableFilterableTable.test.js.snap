// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LovableFilterableTable user enters search query should filter items (with serializer!) 1`] = `
<tbody>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Bitcoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BTC
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      4069.35
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -0.21
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -1.09
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -3.36
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Bitcoin Cash
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BCH
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      619.897
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -4.33
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -20.3
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      104.19
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Litecoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      LTC
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      46.8575
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.13
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      3.86
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      2.11
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Binance Coin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BNB
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      2.59805
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      2.54
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      8.77
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      93.52
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Bytecoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BCN
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.00118761
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -0.35
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -2.26
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -11.95
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
</tbody>
`;

exports[`LovableFilterableTable user enters search query should filter items 1`] = `
Array [
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin Cash</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Litecoin</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Binance Coin</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Bytecoin</td>",
]
`;

exports[`LovableFilterableTable user enters search query should render a subset of matching \`items\` 1`] = `"<tbody><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BTC</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">4069.35</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">-0.21</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-1.09</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">-3.36</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin Cash</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BCH</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">619.897</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">-4.33</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-20.3</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">104.19</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Litecoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">LTC</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">46.8575</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">0.13</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">3.86</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">2.11</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Binance Coin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BNB</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">2.59805</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">2.54</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">8.77</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">93.52</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Bytecoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BCN</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">0.00118761</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">-0.35</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-2.26</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">-11.95</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr></tbody>"`;

exports[`LovableFilterableTable user enters search query testing out snapshots 1`] = `
Object {
  "24h_volume_usd": "2481990000.0",
  "available_supply": "16518812.0",
  "id": "bitcoin",
  "isLoved": false,
  "last_updated": "1503319469",
  "market_cap_usd": "67220827612.0",
  "name": "Bitcoin",
  "percent_change_1h": "-0.21",
  "percent_change_24h": "-1.09",
  "percent_change_7d": "-3.36",
  "price_btc": "1.0",
  "price_usd": "4069.35",
  "rank": "1",
  "symbol": "BTC",
  "total_supply": "16518812.0",
}
`;
